pkg/cache/v3/linear.go

```diff
@@ -114,60 +114,116 @@ func NewLinearCache(typeURL string, opts ...LinearCacheOption) *LinearCache {
+				removedResources = append(removedResources, resourceName)
 			}
 		}
 	} else {
-		// Non-wildcard request for other types. Only return stale resources
-		resources = make([]types.ResourceWithTTL, 0, len(staleResources))
-		for _, name := range staleResources {
-			resource := cache.resources[name]
-			if resource != nil {
-				resources = append(resources, types.ResourceWithTTL{Resource: resource})
+		for resourceName := range watch.subscription.SubscribedResources() {
+			version, exists := cache.versionVector[resourceName]
+			knownVersion, known := knownVersions[resourceName]
+			if !exists {
+				if known {
+					// This resource was removed and is no longer known. If the type requires full state
+					// we need to return to make it aware
+					removedResources = append(removedResources, resourceName)
+				}
+				continue
+			}
+
+			if !known {
+				// This resource is not yet known by the client (added or the client has a stale view)
+				changedResources = append(changedResources, resourceName)
+			} else if knownVersion != version {
+				// Outdated version
+				changedResources = append(changedResources, resourceName)
 			}
 		}
 	}
-	watch.Response <- &RawResponse{
-		Request:   watch.Request,
-		Resources: resources,
-		Version:   cache.getVersion(),
-		Ctx:       context.Background(),
+
+	if len(changedResources) == 0 && len(removedResources) == 0 {
+		// Nothing changed
+		return nil
+	}
+
+	returnedVersions := make(map[string]string, len(watch.subscription.ReturnedResources()))
+	// Clone the current returned versions. The cache should not alter the subscription
+	for resourceName, version := range watch.subscription.ReturnedResources() {
+		returnedVersions[resourceName] = version
+	}
+
+	// Depending on the type, the response will only include changed resources or all of them
+	cacheVersion := cache.getVersion()
+	var resources []types.ResourceWithTTL
+	if ResourceRequiresFullStateInSotw(cache.typeURL) {
+		resources = make([]types.ResourceWithTTL, 0, len(cache.resources))
+		for resourceName, resource := range cache.resources {
+			resources = append(resources, types.ResourceWithTTL{Resource: resource})
+			returnedVersions[resourceName] = cacheVersion
+		}
+	} else {
+		resources = make([]types.ResourceWithTTL, 0, len(changedResources))
+		for _, resourceName := range changedResources {
+			resources = append(resources, types.ResourceWithTTL{Resource: cache.resources[resourceName]})
+			returnedVersions[resourceName] = cacheVersion
+		}
+	}
+	return &RawResponse{
+		Request:           watch.Request,
+		Resources:         resources,
+		ReturnedResources: returnedVersions,
+		Version:           cacheVersion,
+		Ctx:               context.Background(),
 	}
 }
 
-func (cache *LinearCache) notifyAll(modified map[string]struct{}) {
-	// de-duplicate watches that need to be responded
-	notifyList := make(map[ResponseWatch][]string)
+func (cache *LinearCache) notifyAll(ctx context.Context, modified map[string]struct{}) {
```

Comment:
what's the point of passing a context here?

Reply:
Cleaned up. We return a context in the response interface. In the case of Snapshot cache we do pass in the context from SetSnapshot. Not sure what we want to do here to be honest, but removed it for now